RFC-0026-logging-system.md

```diff
@@ -0,0 +1,481 @@
+# New PyTorch Logging System
+
+## **Summary**
+Create a message logging system for PyTorch with the following requirements:
+
+### Consistency 
+
+* The C++ and Python APIs should match each other as closely as possible.
+
+* All errors, warnings, and other messages generated by PyTorch should be
+  emitted using the the logging system API.
+
+
+### Severity level and message classes
+
+* Offer different message severity levels, including at least the following:
+
+  - **Info**: Emits a message without creating a warning or error. By default,
+    this gets printed to stdout.
+
+  - **Warning**: Emits a message as a warning. If a warning is never caught,
+    it gets printed to stderr by default.
+
+  - **Error**: Emits a message as an error. If an error is never caught, the
+    application will print the error to stderr and quit.
+
+  - TODO: Do we also need a **Fatal** severity for integration with Meta's
+    internal logging system (glog)? A fatal message terminates the program
+
+* Offer different message classes under each severity level.
+
+  - Every message is emitted as an instance of a message class.
+
+  - Each message class has both a C++ class and a Python class, and when a
+    C++ message is propagated to Python, it is converted to its corresponding
+    Python class.
+
+  - Whenever it makes sense, the Python class should be one of the builtin
+    Python error/warning classes. For instance, currently in PyTorch, the C++
+    error class `c10::Error` gets converted to the Python `RuntimeError` class.
+
+* Adding new message classes and severity levels should be easy
+
+### Configurability and filtering
+
+* Ability to turn warnings into errors. This is already possible with the
+  Python `warnings` module filter, but the PyTorch docs should mention it and
+  we should probably have unit tests for it.
+  See [documentation](https://docs.python.org/3/library/warnings.html#the-warnings-filter)
+
+* Settings to disable specific message classes and severity levels
+
+  - TODO: Error classes should never be disableable, right?
```

Comment:
AFAIK there is no good reason anymore for the `python -OO` mode; it strips not only asserts but also docstrings, in an attempt to save non-essential space (which probably was useful 20 years ago).

There's one web framework that still launches with `-OO`, which is a recurring source of issues in numpy/scipy; any manipulation of `__doc__` should be guarded because `__doc__` can be `None`. It's a hassle and fairly pointless.

> So I would vote for no.

me too